code/game/objects/structures/statues.dm

```diff
@@ -293,10 +253,281 @@
+
+/obj/structure/carving_block/proc/reset_target()
+	current_target = null
+	current_preset_type = null
+	target_appearance_with_filters = null
+
+/obj/structure/carving_block/update_overlays()
+	. = ..()
+	if(target_appearance_with_filters)
+		//We're only keeping one instance here that changes in the middle so we have to clone it to avoid managed overlay issues
+		var/mutable_appearance/clone = new(target_appearance_with_filters)
+		. += clone
+
+/obj/structure/carving_block/proc/is_viable_target(atom/movable/target)
+	//Only things on turfs
+	if(!isturf(target.loc))
+		return FALSE
+	//No big icon things
+	var/icon/thing_icon = icon(target.icon, target.icon_state)
+	if(thing_icon.Height() != world.icon_size || thing_icon.Width() != world.icon_size)
+		return FALSE
+	return TRUE
+
+/obj/structure/carving_block/proc/create_statue()
+	if(current_preset_type)
+		var/obj/structure/statue/preset_statue = new current_preset_type(get_turf(src))
+		preset_statue.set_custom_materials(custom_materials)
+		qdel(src)
+	else if(current_target)
+		var/obj/structure/statue/custom/new_statue = new(get_turf(src))
+		new_statue.set_visuals(current_target)
+		new_statue.set_custom_materials(custom_materials)
+		var/mutable_appearance/ma = current_target
+		new_statue.name = "statue of [ma.name]"
+		new_statue.desc = "statue depicting [ma.name]"
+		qdel(src)
+
+/obj/structure/carving_block/proc/set_completion(value)
+	if(!current_target)
+		return
+	if(!target_appearance_with_filters)
+		target_appearance_with_filters = new(current_target)
+		target_appearance_with_filters.appearance_flags |= KEEP_TOGETHER
+		target_appearance_with_filters.filters = filter(type="color",color=greyscale_with_value_bump,space=FILTER_COLOR_HSV)
+	completion = value
+	var/static/icon/white = icon('icons/effects/alphacolors.dmi', "white")
+	switch(value)
+		if(0)
+			//delete uncovered and reset filters
+			filters -= partial_uncover_filter
+			target_appearance_with_filters = null
+		else
+			var/mask_offset = min(world.icon_size,round(completion * world.icon_size))
+			if(partial_uncover_filter)
+				filters -= partial_uncover_filter
+			partial_uncover_filter = filter(type="alpha",icon=white,y=-mask_offset)
+			filters += partial_uncover_filter
+			target_appearance_with_filters.filters = filter(type="alpha",icon=white,y=-mask_offset,flags=MASK_INVERSE)
+	update_icon()
+
+
+/// Returns a list of preset statues carvable from this block depending on the custom materials
+/obj/structure/carving_block/proc/get_possible_statues()
+	. = list()
+	if(!statue_costs)
+		statue_costs = build_statue_cost_table()
+	statue_loop:
+		for(var/statue_path in statue_costs)
+			var/list/carving_cost = statue_costs[statue_path]
+			for(var/required_material in carving_cost)
+				if(!custom_materials[required_material] || custom_materials[required_material] < carving_cost[required_material])
+					continue statue_loop
+			. += statue_path
```

Comment:
```suggestion
	for(var/statue_path in statue_costs)
		var/valid_path = TRUE
		var/list/carving_cost = statue_costs[statue_path]
		for(var/required_material in carving_cost)
			if(!custom_materials[required_material] || custom_materials[required_material] < carving_cost[required_material])
				valid_path = FALSE
				break
		if(!valid_path)
			continue
		. += statue_path
```
goto evil
Double-check if this does what you intended. A bit confused on why you wanted to completely restart the loop each time.

Reply:
`continue label` preservers the loop counter, but yeah guardian var is better here.